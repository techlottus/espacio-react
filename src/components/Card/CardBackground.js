import React, { useEffect } from 'react';



export const CardBackground = ({data}) => {

    const cardRef = React.createRef();
  
    useEffect(() => {
      cardRef.current.data = {
        title: data.title || '',
        subtitle: data.subtitle || '',
        disabled: data.disabled,
        image: data.image || '',
      };
    })
  
    return (
      <>
        <lottus-card-background ref={cardRef}></lottus-card-background>
      </>
    )
  }