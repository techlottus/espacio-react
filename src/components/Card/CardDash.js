import React, { useEffect } from 'react';


export const CardDash = React.memo(({data,onClick,onBtn}) => {

    const cardRef = React.createRef();
    const buttonRef = React.createRef();
  
    useEffect(() => {
      cardRef.current.data = {
        icon: data.icon || '',
        title: data.title || '',
        disabled: data.disabled,
        width: data.width || '',
        height: data.height || '',
        test: data.test || ''
      };
      buttonRef.current.data = {
        ...data.action
      }
    },[data]) // eslint-disable-line react-hooks/exhaustive-deps

    useEffect(() => {
      buttonRef.current.addEventListener('onClick', onBtn);
      cardRef.current.addEventListener('onClick', onClick);
    },[]) // eslint-disable-line react-hooks/exhaustive-deps
  
    return (
      <>
        <lottus-card-dash style={{width: '100%', height:'100%'}} ref={cardRef}>
          <lottus-button ref={buttonRef} button></lottus-button>
        </lottus-card-dash>
      </>
    )
  })