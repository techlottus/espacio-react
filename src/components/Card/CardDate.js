import React, { useEffect } from 'react';



export const CardDate = ({data}) => {

    const cardRef = React.createRef();
  
    useEffect(() => {
      cardRef.current.data = {
        day: data.day || 0,
        dayName: data.dayName || '',
        title: data.title || '',
        items: data.item || [],
        disabled: data.disabled,
      };
    })
  
    return (
      <>
        <lottus-card-date ref={cardRef}></lottus-card-date>
      </>
    )
  }