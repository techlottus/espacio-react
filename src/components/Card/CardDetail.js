import React, { useEffect } from 'react';



export const CardDeatil = ({data,onCheck,onAction}) => {

    const cardDetailRef = React.createRef();
  
    useEffect(() => {
      cardDetailRef.current.data = {
        id: data.id || '',
        title: data.title || '',
        description: data.description || '',
        price: data.price || 0,
        action: data.action || '',
        checked: data.checked,
        test: data.test || '',
        testAction: data.testAction || ''
      };
    },[data]) // eslint-disable-line react-hooks/exhaustive-deps

    useEffect(() => {
      cardDetailRef.current.addEventListener('onCheck',onCheck);
      cardDetailRef.current.addEventListener('onAction',onAction)
    },[]) // eslint-disable-line react-hooks/exhaustive-deps
  
    return (
      <>
        <lottus-card-detail ref={cardDetailRef}></lottus-card-detail>
      </>
    )
  }