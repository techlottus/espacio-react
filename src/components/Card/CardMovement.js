import React, { useEffect } from "react";



export const CardMovement = ({ data, onBtn, onClick }) => {
  const cardRef = React.createRef();
  const buttonRef = React.createRef();

  useEffect(() => {
    cardRef.current.data = {
      date: data.date || "",
      title: data.title || "",
      color: data.color || "",
      text: data.text || "",
      price: data.price || 0,
      rows: data.rows || [],
    };
    if (data.action !== null) {
      buttonRef.current.data = {
        type: data.action.type || "primary",
        title: data.action.title || "",
        size: data.action.size || "small",
        icon: data.action.icon || "",
        lyIcon: data.action.lyIcon,
        disabled: data.action.disabled,
        isExpand: data.action.isExpand,
      };
    }
  }, [data]); // eslint-disable-line react-hooks/exhaustive-deps

  useEffect(() => {
    if(data.action !== null){
      buttonRef.current.addEventListener("onClick", onBtn);
    }
    cardRef.current.addEventListener("onClick", onClick);
  }, []); // eslint-disable-line react-hooks/exhaustive-deps

  return (
    <>
      <lottus-card-movement ref={cardRef}>
        {data.action !== null ? (
          <div btnmovement="">
            <lottus-button ref={buttonRef}></lottus-button>
          </div>
        ) : null}
      </lottus-card-movement>
    </>
  );
};
