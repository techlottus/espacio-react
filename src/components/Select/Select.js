import React, { useEffect } from 'react';



export const Select = React.memo(({data, options,onClick}) => {

    const selectRef = React.createRef();
  
    useEffect(() => {
      selectRef.current.data = {
        textDefault: data.textDefault || '',
        disabled: data.disabled,
        reset: data.reset,
        icon: data.icon || '',
        zindexOptions: 100,
        isLabel: data.isLabel || false
      };
    },[data]) // eslint-disable-line react-hooks/exhaustive-deps 

    useEffect(() => {
      selectRef.current.options = options || null
    },[options]); // eslint-disable-line react-hooks/exhaustive-deps

    useEffect(() => {
      selectRef.current.addEventListener('onClick', onClick);
    },[]) // eslint-disable-line react-hooks/exhaustive-deps
  
    return (
      <>
        <lottus-select ref={selectRef}></lottus-select>
      </>
    )
})