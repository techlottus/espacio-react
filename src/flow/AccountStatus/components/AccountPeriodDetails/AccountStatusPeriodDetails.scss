@import "../../../../App.scss";

.accountdetails-container {
  font-family: "Nunito", sans-serif;
  background: map-get($palette-neutral, n00);
  padding: 0 pixeltorem(14px);
  @media only screen and (min-width: map-get($grid-breakpoints, m)) and (max-width: map-get($grid-breakpoints, xl)) {
    padding: 0;
  }
  h2 {
    &.headingdetails{
      font-size: pixeltorem(18px);
      font-weight: 700;
    }
  }

  p {
    &.textdetails{
      font-size: pixeltorem(14px);
      padding: pixeltorem(10px) 0;
      line-height: pixeltorem(19px);
    }
  }

  .detail-bullets {
    display: flex;
    padding: pixeltorem(10px) 0 pixeltorem(41px) 0;

    p {
      font-size: pixeltorem(14px);
      display: flex;
      padding: 0;
      padding-right: pixeltorem(24px);
      &.org-p {
        color: #f02b00;
      }
      &.gr-p {
        color: #8dc881;
      }
      &.pr-p {
        color: #682e8d;
      }
      span {
        height: pixeltorem(14px);
        width: pixeltorem(14px);
        border-radius: 50%;
        margin-right: pixeltorem(8px);
        display: inline-block;
        &.org-bullet {
          background-color: #f02b00;
        }
        &.gr-bullet {
          background-color: #8dc881;
        }
        &.pr-bullet {
          background-color: #682e8d;
        }
      }
    }
  }
  .accountdetails-selects {
    width: 100%;
    .select-details{
      padding-bottom: pixeltorem(16px);
      p{
        &.textprogram{
          font-size: pixeltorem(14px);
          font-weight: 700;
          padding-bottom: pixeltorem(8px)
        }
      }

      @media only screen and (min-width: map-get($grid-breakpoints, m)) {
        width: calc(50% - #{pixeltorem(12px)});
        padding-right: pixeltorem(12px);
        &:nth-child(2) {
          padding-right: 0;
          padding-left: pixeltorem(12px);
  
        }
      }
    }

    @media only screen and (min-width: map-get($grid-breakpoints, m)) {
      padding-top: pixeltorem(24px);
      display: flex;
    }
  }

  .accountdetails-period {
    padding-top: pixeltorem(48px);
    display: flex;
    justify-content: space-between;
    align-items: center;
    p {
      &.textperiod{
        padding: pixeltorem(9px) 0;
        &.detail-amount {
          padding: 0;
          font-weight: 700;
          font-size: pixeltorem(18px);
        }
      }
    }

    .calendar {
      padding: pixeltorem(9px) 0;
      display: flex;
      align-items: center;
      span {
        padding-right: pixeltorem(11px);
      }
    }
  }

  .accountdetails-movements {
    padding: pixeltorem(36px) 0;

    .cardmovement{
      padding: pixeltorem(8px) 0;
    }
  }

  .accountperioddetailserror {
    padding-top: pixeltorem(16px);
    padding-bottom: pixeltorem(32px);
  }
}
