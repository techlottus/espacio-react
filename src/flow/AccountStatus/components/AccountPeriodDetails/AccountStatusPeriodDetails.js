import React, { useEffect, useState } from "react";
import { Select } from "../../../../components/Select/Select";
import { Label } from "../../../../components/Label/Label";
import { CardMovement } from "../../../../components/Card/CardMovement";
import { getMovementsService } from "../../../../middlewares/accountStatusMiddleware";
import { convertMoney } from "../../../../helpers/formatCurrency";
import {
  selectProgram,
  selectPeriod,
  labelInit,
  feedbackErrorAccountStatus,
} from "../../../../constants/AccountStatus.constant";
import "./AccountStatusPeriodDetails.scss";
import { useDispatch, useSelector } from "react-redux";
import { useHistory } from "react-router";
import { convertPeriod } from "../../../../helpers/accountStatusTypes";
import { formatDate } from "../../../../helpers/formatDate";
import { setConceptDefault } from "../../../../actions/paymentAction";
import { formatString } from "../../../../helpers/formatString";
import { Feedback } from "../../../../components/Feedback/Feedback";
import { typesRequestErrors } from "../../../../constants/error.constant";
import { sendInfoTM } from "../../../../tagging/services/sendInfoTagManager";
import { btnPaymentAccountStatusTag, selectPeriodAccountStatusTag, selectProgramAccountStatusTag } from "../../../../tagging/flows/accountStatusTag";

const AccountStatusPeriodDetails = () => {
  const dispatch = useDispatch();
  const history = useHistory();
  const accountStatusStore = useSelector((state) => state.accountStatus);
  const [errorPrograms, setErrorPrograms] = useState({
    isError: false,
    msg: null,
  });

  const [errorMovements, setErrorMovements] = useState({
    isError: false,
    msg: null,
  });

  const [movements, setMovements] = useState(null);

  const [filters, setFilters] = useState({
    program: null,
    period: null,
  });

  const [selectPrograms, setSelectPrograms] = useState({
    valid: false,
    index: null,
  });

  const [selectPeriods, setSelectPeriods] = useState({
    valid: false,
    index: null,
  });

  const [programs, setPrograms] = useState([]);

  const [periods, setPeriods] = useState([]);

  const changePrograms = (value) => {
    if (programs && programs.length > 0) {
      const index = programs.findIndex((e) => e.value === value);
      const periodsList = accountStatusStore.programs[index].periods.map(
        (period, index) => {
          return {
            text: `${period.periodDescription}/${formatDate(
              period.startDate,
              period.endDate
            )}`,
            value: period.periodCode,
            active: index === 0,
          };
        }
      );
      setFilters((state) => {
        return {
          ...state,
          period: accountStatusStore.programs[index].periods[0].periodCode,
          program: selectPrograms.index,
        };
      });
      setPeriods(periodsList);
    }
  };

  useEffect(() => {
    if (filters.program !== null && filters.period !== null) {
      setMovements(null);
      dispatch(getMovementsService(filters.period, filters.program));
    }
  }, [filters.program, filters.period]); // eslint-disable-line react-hooks/exhaustive-deps

  useEffect(() => {
    if (selectPrograms.valid) {
      changePrograms(selectPrograms.index);
      setSelectPrograms({
        valid: false,
        index: null,
      });
    }
  }, [selectPrograms]); // eslint-disable-line react-hooks/exhaustive-deps

  useEffect(() => {
    if (selectPeriods.valid) {
      setFilters((state) => {
        return {
          ...state,
          period: selectPeriods.index,
        };
      });
      setSelectPeriods({
        valid: false,
        index: null,
      });
    }
  }, [selectPeriods]);

  useEffect(() => {
    if (accountStatusStore.programs !== null) {
      const programsList = accountStatusStore.programs.map((program, index) => {
        return {
          value: program.programCode || "",
          text: formatString(program.programDescription || ""),
          active: index === 0,
        };
      });
      setPrograms(programsList);
      const periodsList = accountStatusStore?.programs[0].periods.map(
        (period, index) => {
          return {
            text: `${period.periodDescription}/${formatDate(
              period.startDate,
              period.endDate
            )}`,
            value: period.periodCode,
            active: index === 0,
          };
        }
      );
      setFilters((state) => {
        return {
          ...state,
          period: accountStatusStore.programs[0].periods[0].periodCode,
          program: programsList[0].value,
        };
      });
      setPeriods(periodsList);
    }
  }, [accountStatusStore.programs]);

  useEffect(() => {
    if (accountStatusStore.movements !== null) {
      setMovements(accountStatusStore.movements);
    }
  }, [accountStatusStore.movements]);

  useEffect(() => {
    if (
      accountStatusStore.errors[typesRequestErrors.getProgramsAccountStatus]
        .isError
    ) {
      setErrorPrograms({
        isError: true,
        msg: accountStatusStore.errors[
          typesRequestErrors.getProgramsAccountStatus
        ].msg,
      });
    }

    if (
      accountStatusStore.errors[
        typesRequestErrors.getMovementsOfPeriosAccountAccountStatus
      ].isError
    ) {
      setErrorMovements({
        isError: true,
        msg: accountStatusStore.errors[
          typesRequestErrors.getMovementsOfPeriosAccountAccountStatus
        ].msg,
      });
    }

    return () => {
      setErrorPrograms({
        isError: false,
        msg: null
      });
      setErrorMovements({
        isError: false,
        msg: null
      })
    }
  }, [accountStatusStore.errors]);

  const handleClickPayment = (conceptCode) => {
    dispatch(setConceptDefault(conceptCode));
    history.push("/payment");
  };

  const handleSelects = () => {
    return (
      <div className="accountdetails-selects">
        <div className="select-details">
          <p className="textprogram">Programa</p>
          <Select
            data={selectProgram.data}
            options={programs}
            onClick={(e) => {
              sendInfoTM(window,selectProgramAccountStatusTag);
              setSelectPrograms({
                valid: true,
                index: e.detail,
              });
            }}
          />
        </div>
        <div className="select-details">
          <p className="textprogram">Periodo</p>
          <Select
            data={selectPeriod.data}
            options={periods}
            onClick={(e) => {
              sendInfoTM(window,selectPeriodAccountStatusTag);
              setSelectPeriods({
                valid: true,
                index: e.detail,
              });
            }}
          />
        </div>
      </div>
    );
  };

  const handleDetails = () => {
    return (
      <>
        <div className="accountdetails-period">
          <div>
            <h2 className="headingdetails">
              Periodo {(movements && movements.periodCode) || ""}
            </h2>
            <div className="calendar">
              <span className="material-icons icon iconright">
                calendar_today
              </span>
              <p className="textperiod">
                {movements && movements.startDate
                  ? formatDate(movements.startDate, movements.endDate)
                  : ""}
              </p>
            </div>
          </div>
          <div>
            <p>Saldo del periodo</p>
            <p className="detail-amount textperiod">
              {convertMoney(
                movements && movements.periodBalance
                  ? movements.periodBalance
                  : 0
              )}
            </p>
          </div>
        </div>
        <div>
          <Label
            data={{
              ...labelInit,
              title:
                movements && movements.periodType
                  ? convertPeriod(movements.periodType)
                  : "",
            }}
          />
        </div>
        <div className="accountdetails-movements">
          <h2 className="headingdetails">Movimientos</h2>
          <div>
            {movements !== null &&
            movements?.cards &&
            movements?.cards.length > 0
              ? movements.cards.map((data, i) => {
                  return (
                    <div className="cardmovement" key={i + "dev"}>
                      <CardMovement
                        key={i}
                        data={data}
                        onBtn={() => {
                          sendInfoTM(window,btnPaymentAccountStatusTag)
                          handleClickPayment(data.code);
                        }}
                      />{" "}
                    </div>
                  );
                })
              : ""}
          </div>
        </div>
      </>
    );
  };

  const handlePrograms = () => {
    return (
      <>
        {handleSelects()}
        {!errorMovements.isError
          ? handleDetails()
          : handleErrorFeedback(errorMovements.msg)}
      </>
    );
  };

  const handleErrorFeedback = (message) => {
    return (
      <div className="accountperioddetailserror">
        <Feedback data={feedbackErrorAccountStatus} text={message} />
      </div>
    );
  };

  return (
    <div className="accountdetails-container">
      <h2 className="headingdetails">Detalles por periodo</h2>
      <p className="textdetails">
        Selecciona el programa y el periodo para ver los detalles de los
        movimientos; se clasifican de la siguiente manera:
      </p>
      <div className="detail-bullets">
        <p className="org-p">
          <span className="org-bullet"></span>Con adeudo
        </p>
        <p className="gr-p">
          <span className="gr-bullet"></span>Pagados
        </p>
        <p className="pr-p">
          <span className="pr-bullet"></span>Por pagar
        </p>
      </div>
      {errorPrograms.isError
        ? handleErrorFeedback(errorPrograms.msg)
        : handlePrograms()}
    </div>
  );
};

export default AccountStatusPeriodDetails;
