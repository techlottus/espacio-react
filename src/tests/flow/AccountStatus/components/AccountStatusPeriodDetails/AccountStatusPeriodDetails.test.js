import React from "react";
import { mount } from "enzyme";
import { Provider, useDispatch } from "react-redux";
import "@testing-library/jest-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import AccountStatusPeriodDetails from "../../../../../flow/AccountStatus/components/AccountPeriodDetails/AccountStatusPeriodDetails";
import { typesRequestErrors } from "../../../../../constants/error.constant";
import { Select } from "../../../../../components/Select/Select";
import { act } from "react-dom/test-utils";
import { filtersMock, formatProgramsMock, movementsMock, programsMock } from "../../../../../fixtures/accountStatusPeriodDetailsFixture";
import { renderHook } from "@testing-library/react-hooks";
import { CardMovement } from "../../../../../components/Card/CardMovement";
import { BrowserRouter } from "react-router-dom";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const historyMock = { push: jest.fn(), location: {}, listen: jest.fn() };

let initState = {
  accountStatus: {
    profile: null,
    movements: null,
    errors: {
      [typesRequestErrors.getAccountStatus]: {
        isError: false,
        msg: null,
      },
      [typesRequestErrors.getAccountMovements]: {
        isError: false,
        msg: null,
      },
    },
    programs: null,
    success: {
      isSuccess: false,
      msg: null,
    },
    debitPeriods: null,
  },
};


let store = mockStore(initState);

const { result } = renderHook(() => useDispatch())

const handleMount = (initState) => {
  store = mockStore(initState);  
  return mount(
    <Provider store={store}>
      <BrowserRouter>
        <AccountStatusPeriodDetails history={historyMock} />
      </BrowserRouter>
    </Provider>
  );
};

describe("Test <AccountStatusPeriodDetails />", () => {
  beforeEach(() => {});

  test("should init component", () => {
    const wrapper = handleMount(initState);
    expect(wrapper).toMatchSnapshot();
  });

  test("should click select program", () => {
    const wrapper = handleMount(initState);
    act(() => {
      wrapper.find(Select).first().props().onClick({   
        detail: 0,
      });
    });
    wrapper.update();
  });

  test("should click select period", () => {
    initState = {
      ...initState,
      accountStatus: {
        ...initState.accountStatus,
        programs: programsMock
      }
      
    };
    const wrapper = handleMount(initState);
    expect(wrapper.find(Select).exists()).toBe(true);
    act(() => {
      wrapper.find(Select).last().props().onClick({
        detail: 0,
      });
    });
    wrapper.update();
  });

  test("should valid programs and periods", () => {
    const wrapper = handleMount(initState);
    expect(wrapper.find(Select).first().props().options).toEqual(formatProgramsMock);
    act(() => {
      wrapper.find(Select).first().props().onClick({
        detail: 'LEINSC334'
      })
      wrapper.update();
    });
    expect(wrapper.find(Select).last().props().options).toEqual(filtersMock);
    act(() => {
      wrapper.find(Select).last().props().onClick({
        detail: '202042'
      })
      wrapper.update();
    });
    
  });

  test("should valid movements and onBtn", () => {
    initState = {
      ...initState,
      accountStatus: {
        ...initState.accountStatus,
        programs: null,
        movements: movementsMock
      }
      
    };
    const wrapper = handleMount(initState);
    expect(wrapper.find(CardMovement).exists()).toBe(true);
    act(() => {
      wrapper.find(CardMovement).first().props().onBtn({})
    });
    wrapper.update();
  })
});