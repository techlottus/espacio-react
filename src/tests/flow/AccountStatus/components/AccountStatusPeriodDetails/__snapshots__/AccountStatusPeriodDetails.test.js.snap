// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <AccountStatusPeriodDetails /> should init component 1`] = `
<div
  className="accountdetails-container"
>
  <h2
    className="headingdetails"
  >
    Detalles por periodo
  </h2>
  <p
    className="textdetails"
  >
    Selecciona el programa y el periodo para ver los detalles de los movimientos; se clasifican de la siguiente manera:
  </p>
  <div
    className="detail-bullets"
  >
    <p
      className="org-p"
    >
      <span
        className="org-bullet"
      />
      Con adeudo
    </p>
    <p
      className="gr-p"
    >
      <span
        className="gr-bullet"
      />
      Pagados
    </p>
    <p
      className="pr-p"
    >
      <span
        className="pr-bullet"
      />
      Por pagar
    </p>
  </div>
  <div
    className="accountdetails-selects"
  >
    <div
      className="select-details"
    >
      <p
        className="textprogram"
      >
        Programa
      </p>
      <lottus-select />
    </div>
    <div
      className="select-details"
    >
      <p
        className="textprogram"
      >
        Periodo
      </p>
      <lottus-select />
    </div>
  </div>
  <div
    className="accountdetails-period"
  >
    <div>
      <h2
        className="headingdetails"
      >
        Periodo 
      </h2>
      <div
        className="calendar"
      >
        <span
          className="material-icons icon iconright"
        >
          calendar_today
        </span>
        <p
          className="textperiod"
        />
      </div>
    </div>
    <div>
      <p>
        Saldo del periodo
      </p>
      <p
        className="detail-amount textperiod"
      >
        $0.00
      </p>
    </div>
  </div>
  <div>
    <lottus-label />
  </div>
  <div
    className="accountdetails-movements"
  >
    <h2
      className="headingdetails"
    >
      Movimientos
    </h2>
    <div />
  </div>
</div>
`;
